README
Magick -- a system of IRC services for IRC networks
-------------------------------------------------------

Magick is copyright (c) 1996-1997 Preston A. Elder.  Magick may be
freely redistributed; see the GNU General Public License (in the file
"COPYING") for details.


1. INTRODUCTION

     Magick is a system of services (as the name implies) to be used
with Internet Relay Chat networks.  Magick provides for definitive
nickname and channel ownership, as well as the ability to send messages
("memos") to offline users, and gives IRC operators considerably more
control over the network, enabling them to change modes in any channel and
place network-wide bans, among other things.
     Magick runs as a server on an IRC network, and is designed to use
the RFC 1459 IRC protocol, with some additions discussed at the end of this
document.
     Also, Magick will not work with the reference implementation of RFC
1459 (irc2.x); an extension to the code is required to allow Services to
change channel modes without requiring channel operator status.  The DALnet
extension to ircd (which can be found at ftp.dal.net:/pub/dalnet/server) is
one implementation with the necessary extension.
     *** Note as of version 2.2.22:  Basic support for non-DALnet servers
is present.  However, it has not been tested; problem and success reports
both appreciated.  In particular, I'd like to know of any equivalents to
the IRC commands (AKILL, RAKILL, GLOBOPS, GOPER) discussed at the bottom of
this file.


2. CONFIGURATION, COMPILATION AND INSTALLATION

     In order to compile Magick, you'll need a Bourne shell or
compatible shell (such as GNU bash), GNU make or a compatible make (which
needs to support the "include" directive), and an ANSI C compiler (gcc
recommended).
     Before trying to compile Magick, there are some manual configuration
tasks that need to be done: run the "configure" script, edit config.h, and
edit the top section of the Makefile.
     The "configure" script will try to learn how things work on your
system, and set up appropriate Makefile and C source definitions.  It will
also ask you a few questions about how you want Services set up.  (It will
ask all of the questions before doing any of the automated tests.)  If you
get an error while running the script, get bash (if you don't already have
it) and run "bash configure".  IMPORTANT NOTE:  Make sure you select the
correct IRC server style!  If Magick compiles correctly but you only get
messages like "Sorry, registration failed", the most likely cause is that
you chose the wrong IRC server style.
     config.h contains all of the basic definitions for Magick -- uplink
server, data directory, etc.  Most of these can be changed with
command-line options (see below), but you may find it more convenient to
set them as defaults.
     The Makefile has a section at the top allowing you to configure
Magick general mode of operation.  
     Finally, there's one definition in operserv.c (which really should be
moved out, I know) which you'll want to edit: s_GlobalNoticer, which sends
out global notices when OperServ's GLOBAL command is used.  It defaults to
"DarkerNet", which is probably not appropriate for your network. ;^)
     Once these steps are done, you should be able to compile Magick
with little trouble.  If Magick fails to compile on your system, or if
you have to tweak configure's output to get it to work, let me know what
went wrong so I can try and incorporate a fix into the configure script
or the main code.
     Once Magick is compiled, type "make install" to copy the programs
files to their destinations.  If you are installing Magick for the first
time, you also need to run "make install-data" to copy the data files to
the destination data directory.  Care should be taken that only authorized
people have access to these files; passwords are NOT encrypted at present,
so unauthorized access to the files could be a big problem!

     NOTE:  The HelpServ data files (data/helpfiles/*) are set up for the
DarkerNet IRC network; you'll want to change them before starting up
Magick on your own network.
     NOTE 2:  Magick databases are endian-specific; that is, databases
created on a little-endian machine won't work on a big-endian one and
vice-versa (except in the degenerate case of an empty database).
Currently, this means that Magick can't be moved from or to an Intel x86
machine without manually converting the databases, until Intel gets their
act together and starts making big-endian processors like the rest of the
world.


3. OPERATION

     This section does not detail the operation of the individual pieces of
Magick; that information can be found in the online help files
("/msg *Serv help" or read lib/*serv/index).  It only describes how to
start Magick itself.
     Normally, Magick can be run simply by invoking the "magick"
executable.  Magick will then use the defaults specified in the config.h
file, and connect to the specified uplink server.  Alternatively, any of
the following command-line options can be specified to change the default
values:
	-remote server[:port]	Connect to the specified server
	-name servername	Our server name (e.g. hell.darker.net)
	-desc string		Description of us (e.g. DarkerNet Services)
	-user username		Username for Services' nicks (e.g. services)
	-host hostname		Hostname for Services' nicks (e.g. darker.net)
	-dir directory		Directory containing Services' data files
				    (e.g. /usr/local/lib/services)
	-log filename		Services log filename (e.g. magick.log)
	-tz timezone		Name of Services' timezone (e.g. EST)
	-update secs		How often to update databases, in seconds
	-debug			Enable debugging mode--more info sent to log
	-relink secs		Specify seconds to re-attempt link if SQUIT
	-norelink		Die on SQUIT or kill -1
	-level level		Specify services level (>1 = backup)

     Upon starting, Magick will parse its command-line parameters, open
its logfile, then (assuming all went well) detach itself and run in the
background.
     If Magick is compiled with SERVICES_LEVEL > 1, it can serve as a
"backup" to the full version of Services.  A "full" version of Services
(of level 1) will automatically reintroduce its pseudo-clients on nick
collide, while a "backup" Services will not, thus allowing full Services
to be brought up at any time without disrupting the network (and without
having to take backup Services down beforehand).
     Two additional programs available in addition to the main executable:
"listnicks" and "listchans", both installed as hard links to the main
executable.  The programs will list all registered nicknames and channels,
respectively; or, if given the -c option, will display the number of
registered nicknames or channels.
     If you have anykind of "relink" mode active (ie. you didnt undef the
SERVER_RELINK define or set it to -1, or didnt use the -norelink flag),
only an OperServ QUIT or SHUTDOWN or a kill -9 will stop Magick.  A SQUIT
will just cause relink (after period specified with SERVER_RELINK or the
-relink flag), as will a kill -1 (both causing the databases to be reloaded).

4. OVERVIEW OF SERVICES CLIENTS

     This is a brief introduction to the various clients available from
Magick.  All *Serv clients can be /msg'd with "help" for a general
introduction.
     NickServ is the nickname server; it allows users to register and
control nicknames, and will (at the user's choice) /kill any unauthorized
user who tries to use that nick after a warning.
     ChanServ is the channel server; as NickServ, it allows users to
register and control channels.  There is a much wider array of controls
available via ChanServ than NickServ, since there are considerably more
features available.  These include automatic mode setting, topic
retention (active by default, this will cause the channel's topic to be
saved while the channel is empty and restored when a user joins it again),
and automatic opping (and kicking) of selected users.
     MemoServ allows users to send short messages to other users, which can
be stored and retrieved at the recipient's convenience.
     HelpServ is Services' help system; it will, on request, send a help
text to a user.  Other Services clients also call on HelpServ for their
own help functions.  (So, for example, "/msg ChanServ help register" is the
same as "/msg HelpServ chanserv register".
     IrcIIHelp is HelpServ under another name, and allows ircII users to
simply type "/help <topic>" to get help on the ircII client.
     OperServ provides services to IRC operators, including the ability to
change the mode of any channel, kick any user from a channel, and add and
remove network-wide bans (similar to classic K:lines, but applying to all
servers on the network).  A certain set of operators can be defined (in
config.h) as Services operators, and can perform additional functions, such
as manually updating Magick' databases or shutting Magick down.
(Note that Magick will not recognize a user as a Services operator unless
that user is in the SERVICES_OPS definition, has identified with NickServ and
is oper'd up.)  Obviously, all these functions should be used with care.
     DevNull is just like its Unix equivalent /dev/null: it ignores
anything sent to it.  It can be removed without affecting the rest of
Magick in any way.
     Death is the global noticer: when Magick is instructed to send a
notice to all clients on the network, this nickname sends the message.
Obviously, you should change it to the name of your network (or some other
useful name)!  See operserv.c for the name definition.


5. IRC PROTOCOL ADDITIONS

Magick has not been tested on bare (non-TS8) irc2.x servers; please
report any problems.

The following commands, not defined in RFC 1459, are used by Magick:

AKILL

    Syntax: AKILL <hostmask> <usermask> <reason>

    Adds an AutoKill to the network; this is like a K:line, but is
    propogated to all servers.  Any user matching the usermask and
    hostmask will not be allowed to connect.

    Example:  :hell.darker.net AKILL *.lame.com lamer :Flooding
        Disallows any client matching "lamer@*.lame.com" from connecting
        to the network.

RAKILL

    Syntax: RAKILL <hostmask> <usermask>

    Removes an AutoKill line from all servers.

    Example:  :hell.darker.net RAKILL *.lame.com lamer
        Removes the AutoKill described in the previous example.

GLOBOPS

    Syntax: GLOBOPS <message>

    Sends a message to all users with user mode +og.

    Example:  :PreZ GLOBOPS :Watch out for flooders from *.lame.com.

GOPER

    Syntax: GOPER <message>

    Sends a message to all IRC operators, regardless of other user modes.

    Example:  :hell.darker.net GOPER :WARNING -- clones detected from
                      ppp1-9.lame.com


ALSO for recent DALnet ircd's (4.4.15+) the specialised services code is
used to enable greater control via. services, and enable greater intergration
including things like setting mode +r for registered nicks, etc.

6. REACHING THE AUTHOR

     I can be reached via E-mail at prez@antisocial.com.  Please feel
free to send comments, suggestions, problem reports, diffs, or whatever,
though you may not receive any reply.  Please do NOT ask for or expect
direct online help, as I am quite busy and cannot spare the time to set up
Magick myself on every system where it will be used.  If you do ask,
expect to be abruptly dismissed or ignored.

     If you are on the DarkerNet or DALnet IRC networks and want to leave me
a memo via MemoServ rather than sending E-mail, please include your problem
or suggestion _in the memo_.  Don't just say "I need to talk to you"; such
memos will be ignored.


Changes
Magick I (PreZ <prez@antisocial.com>)
--------

Renamed to this to avoid conflicting version numbers, etc and finally
rid myself of all the excess EsperNet baggage.

1997/12/15 .1	Optimized code and implemented hash tables (if you dont know,
		    dont worry - just take it as a good thing!).
		Added some of Andy Church's services routines (overflow stuff,
		    some socket stuff, and the configure script mainly).
		Removed upaccess - people should have used it, or not need it.
		Added MemoServ OPERSEND.
		Fixed user/oper counting when SQUIT.
		Fixed ghost command and Channel/Nick expiry bugs.
		Added Users on Servers breakdown (OperServ)
		OFF/ON commands now close/open logs aswell.
1997/12/08 .0	Forgot the new services.h for Services 3.0.9, included here.
		Veriable ChanServ levels implemented (finally).
		Added deletion of user from access lists, and all
		    a user's channels when a user is deleted.
		Moved names for services to config.h
		Magick know more about their own actions (dont laugh!)
		IrcOP's dont trigger clone protection (are ignored).
		Added ability to use PRIVMSG instead of NOTICE.
		Added ChanServ JOIN ability (some like it - I dont).
		Added ChanServ REVENGE module, for protecting userlist.
		AKILL and CLONE may now be removed by number.


Version 3.0 (PreZ <prez@antisocial.com>)
-----------

1997/11/30 .9	Little maintinance (checked all various compiles, etc).
		Fixed MAJOR bug with list-duplication on re-connect.
		Minor bugfixes and adds (OFF reasons, pidfile fix, etc)
1997/11/26 .8	Changed the way CLONE detection works (MAJORLY!)
		Added "CLONE" for allowing more than "Default max clones"
		Fixed bug that didnt reset "current online" lists.
		Added "is already there" and "kill all matching" to akill
		Added chanserv Auto-Unban for when ppls are banned on a
		    channel by someone equal to or lower than them.
		Capped the chanserv levels (to stop the "10000" bug).
		Fixed the chanserv UNBAN option so it does ALL bans.
		Added 'bounce bans on link' for def_level[2]+ (10 usually)
1997/11/20 .7	Fixed a few bugs (nickserv IDENT, Globals and Mass -v's)
		Removed SKELETON defines - obsolite by Pick 'n' Choose.
		Removed test/ directory - was old, outdated and useless!
		Added (optional) mask restrictions to USERLIST and CHANLIST.
		AKILL mask's must be >7 chars in length (stop *@*.com).
		Added Services Restart/Reconnect - see config.h for details.
		Added Services Leveling to replace READONLY - better way of
		    implementing backup services versions, etc.
1997/11/10 .6	Fixed up the "Pick 'n' Choose" section of config.h so it
		    now actually makes the end result smaller :)
		Added ChanServ and NickServ SUSPEND/UNSUSPEND functions
		AKILL's will now expire after a time specified in config.h
		Added PAKILL - non-expiring AKILLS settable by SOP's
		Channel/Nick memos will get erased with the channel/nick
		Added ON/OFF so u can enable/disable services without QUIT
		ON, OFF, QUIT, SHUTDOWN and RAW are passworded (config.h)
1997/11/05 .5	Implemented DAL4.4.15+ Services Support, Including:
		    Killing users without a Killed (IrcOp (reason))
		    Server and Nick quarentining via. OperServ
		    Setting of nick's modes with MODE command.
		    CHANGE nick instead of KILL it on nick protection
		SOP's get help for their commands in /msg OperServ HELP
		Memo Forwarding To/From Nick/Channel (in any combo)
		Message to users when they /oper (services.omsg)
		Made use of the till now unused def_access in chanserv.c
1997/10/30 .4	Added support for DAL4.4.15+ servers.
		Moved up all chanserv levels by 5 - 5 is now auto-voice.
		Added /msg operserv listsops (FINALLY!)
		SOPS now need to be /oper'd to be SOPS (oops)
		Oper override of chanserv now #definable (in config.h)
1997/10/22 .3	Fixed a bug in ignore (was stuffing ACCESS list)
		Added OperServ Mode #channel returning current modes
		Opers can override chanserv op/deop/invite/unban
		SOPs can override chanserv clear
1997/10/21 .2	Added READ ALL to MemoServ (channel and nick)
		Added ability for oper to op, deop, invite, unban (ChanServ)
		Added ability for SOP to use clear (ChanServ)
		Added hard-coded /msg HelpServ Credits (so sue me).
		Added memo IGNORE list to nickserv.
		Fixed the nickserv SET IRCOP function
1997/09/15 .1	Only Lvl 20 or sender can delete channel news
		Fixed bug in NickServ that Mis-Saved URL/NICK fields
1997/09/15 .0	Services now respawn if /kill'd
		Message to users when logon added (services.msg)
		Disabled @* (and @*.*, etc) AKILLS except for SOPS
		Disabled @* (and @*.*, etc) AKICKS except for Level 20's
		Created FILE_VERSION 3 (autoconverts from 2)
		Added NickServ URL and EMAIL fields (v3)
		Added ChanServ URL field (v3)
		Added NEWS system (Basically Channel MEMO's)
		Added IRC Operator NickServ flag (no nick expiry)
		Added ChanServ level 20 (can set all but PASS and FOUNDER)
		Added CURRENT TIME to info displays

Version 2.2 (Alcan <achurch@dragonfire.net>
-----------

1997/08/23 .26	Fixed a minor bug in the configure script which caused
		    compilation errors on systems without strerror().
1997/08/21 .25	Fixed a stupid bug in introduce_user() w.r.t. classic ircd.
1997/08/08 .24	Kludged around duplicate QUIT syndrome of Undernet ircd.
1997/08/06	Added multi-server support to introduce_user().
1997/07/24 .23	Fixed user leak caused by autokills.
1997/07/20 .22	Fixed crashes in ChanServ CLEAR {OPS|VOICES}.
1997/07/19	Added ALL option to OperServ STATS command to display
		    Services memory usage.
1997/07/??	Added initial support for different IRC servers.
1997/07/15	Removed remaining occurrences of #define'd parameters
		    (SERVER_NAME, etc.) and replaced them with runtime
		    variables.
1997/07/15	Added "Topic set by" field to ChanServ INFO command.
1997/07/07 .21	Fixed bug in ChanServ CLEAR USERS command causing crashes.
1997/07/02 .20	Removed last vestiges of domain dependence (global notice
		    and MOTD).
1997/07/02	Made IRC operators and Services ops immune from ChanServ

em.h
/*
 * Source copyright Jag Software Limited of England.  Modifications may
 * be made as long as this and other copyright notices (and or messages)
 * are retained in this file.
 *
 * (Why not do the sensible thing and email ME with the source code changes
 *  you would like to implement?)
 *
 *
 * $Source: /home/cvs/magick/Magick-I/em.h,v $
 *
 * $Log: em.h,v $
 * Revision 4.0.0.1  1998/10/17 13:04:53  root
 *
 *
 * Revision 1.4.2.1  1998/10/17 02:28:03  root
 *
 *
 * Revision 1.5  1997/06/29 22:32:15  dev
 * Added g_newer handling, Some minor bug fixes and
 * Mods to enable compilation on more Unix platforms.
 *
 * Revision 1.4  1997/04/03 03:12:57  dev
 * Added defs of invertedLogic(), emStart() and compareLocal()
 * enhancemnets.
 *
 * Revision 1.3  1997/02/11 20:51:41  dev
 * *** empty log message ***
 *
 * Revision 1.2  1996/11/14 02:16:56  dev
 * ftpcwd() returns int, getRegexp takes int
 * Added g_keepAll and g_bytes
 *
 * Revision 1.1  1996/07/10  23:38:13  root
 * Initial revision
 *
 *
*/
static char rcsId_em_h[] = "$Id: em.h,v 4.0.0.1 1998/10/17 13:04:53 root Exp $";



#define FILE_COL 55

#ifndef FALSE
#	define FALSE (0)
#endif

#ifndef TRUE
#	define TRUE (!FALSE)
#endif



#ifndef MAX_PATH
#	define MAX_PATH 255
#endif


#define EQ ==


/*
**	Remote file information.  This is then used to compare against local files
**	etc.
*/
typedef struct
{
	char *fname;	/* File/Dir name */
	char *lnk;		/* Link name (if link) */
	time_t date;	/* File last change date */
	char typ;		/* Type of file */
	long size;		/* Size of file in bytes */
} FileInfo;



#define TYP_DIR	'd'
#define TYP_LNK	'l'
#define TYP_REG	'-'



static char *ftpin( );
static char *ftppwd( );
static void ftpout( );
static void ftppipe( );
static void ftpopen( );
static int ftpcwd( );
static void ftpdup( );
static void ftpget( );


static int cacheRead();
static void cacheWrite();
static char *skip( );
static void ftptimeout( );
static void usage();
static void getRegexp( );
static int ignore();
static void dols_fromlist( );
static int invertedLogic();





#define CHECKERR	if ( errno )											\
					{														\
						fprintf(stderr,"ERROR:%d:%s\n",__LINE__,			\
														strerror(errno));	\
						abort();											\
					}


#define FTP_CONNECT		220
#define FTP_PASSWORD	331
#define FTP_LOGON		230
#define FTP_LOGONFAIL	530
#define FTP_PWD			257
#define FTP_PORTCOMMAND	200
#define FTP_OPEN		150
#define FTP_TRANSFERED	226
#define FTP_CWD			250
#define FTP_FILERROR	550
#define FTP_RECEIVED	432

#define FTP_BYE			221
#define FTP_CLOSE		421
#define FTP_RESET		426




static FileInfo *findFile( );
static void compareLocal();
static void removeDir( );
static time_t getTime( );
static void setTime( );
static int closeSession( );
static int checkforlist( );
static void emStart( );



/*
**	Exit codes
*/
#define EXIT_QUIT		1		/* Quit command actioned */
#define EXIT_ARGS		2		/* Argument error */
#define EXIT_GETDATE	3		/* Getdate() failed.  DATEMSK not set? */


#define EXIT_FTPDOWN	16		/* Ftp just went down */
#define EXIT_SIGNON		17		/* Unable to signon onto remote system */
#define EXIT_REMOTECD	18		/* CD command at remote host failed */

#define EXIT_LOCALCD	32		/* CD command at remote host failed */
#define EXIT_LINK		33		/* Symbolic link failed */
#define EXIT_OPENDIR	34		/* Failed to open a directory */
#define EXIT_STAT		35		/* Stat on a file failed */
#define EXIT_UNLINK		36		/* Failed to unlink */

#define EXIT_FDOPEN		128		/* fdopen() failure */
#define EXIT_PIPE		129		/* Pipe failure */
#define EXIT_FORKEXEC	130		/* fork() or exec() failed */



/*
**	Globals
*/
static int g_debug = FALSE;		/* Enable debug ? */

static int g_timeout = 180;		/* Command timeout. Default is 3 mins */

static unsigned int g_k = 120;	/* 1k should transfer in 2 minutes atleast */

static char g_cache[MAX_PATH+1];/* Cache file */

static int g_preserve = FALSE;	/* Preserve original times on files */

static int g_size = FALSE;		/* Do also compare size to local files, if
								   different then also retrieve file */

static int g_stamp = FALSE;		/* Force timestamps onto local files if
								   newer than remote file? */

static int g_bogus = FALSE;		/* When set to true, files are not
								   retrieved, empty files created instead */

static int g_keepAll = FALSE;	/* By default, non-mirrored files are removed
								   if present on the local system. This flag
  								   prevents local file deletion */

static char g_listfile[1024];	/* Full path spec to the listing file contain
									date/size info for all files on remote 
									system  (eg. ls-lR.gz) */

static char *g_remoteDir;		/* This is the remote directory being mirrored.
								   We need this information for when we use
								   a 'listingfile'.  We'll be searching in the
								   listing file for directory names excluding
								   g_remoteDir from the start.  ie. if we 
								   called em with the remote direcoty as
								   /pub/mirror/slackware, then that directory
								   would be seen as just '' and a subdirectory
								   like /pub/mirror/slackware/abc would be
								   seen as 'abc' */

static int g_openPort = 21;		/* FTP Port to open is defaulted to 21, this
								   can be specified at run-time if a different
								   port is required. */

static unsigned long g_bytes = 0;/*Number of completed bytes transferred */

static char *g_newer = NULL;	/* Only pull files newer than this supplied
								   date. */
